Framework/API/src/ScriptBuilder.cpp

```diff
@@ -265,6 +267,9 @@ const std::string ScriptBuilder::buildPropertyString(
 
       prop = propHistory.name() + opener + propHistory.value() + "'";
     }
+    // Sanitise for a comment #
```

Comment:
Why is this needed? '#' will not open a comment if it is enclosed in quotation marks.

Reply:
It is to do with the way that the recovery script builder constructs the ordered history. It divides each line of workspace history between the command and the time/date stamp and uses `#` as the separating character. 

Currently, a `# ` in the name causes the script builder to separate at the wrong location.

Reply:
Yes, I actually tried it and found the error.

I think we should fix the hacky Python parsing in `OrderWorkspaceHistory` (this is where the error came from for me and it's nature is as you describe, there may be more instances I don't know about) however as `#` is perfectly valid in a Python string.

There is the [`tokenize`](https://docs.python.org/3.5/library/tokenize.html) library which will correctly read lines of Python and separate out comments.

Reply:
@Pasarus - I think that Dan has a point here. Could you try the `tokenize` alternative to this fix, without the changes made here? I think that fix should also reduce the amount of alteration of the original code required.

Reply:
The original code editing was with regards to the way Qt confirmed that all workspaces had been loaded properly, and I simply checked it, to see if a # had been used in the original workspace names and if it had then I would let it continue and through a debug warning that, that had happened, then I would check if it hadn't needed sanitizing normally.

I'm not great with python but I can give it a go.

Reply:
Having re-evaluated I believe it is completely required, in the OrderWorkspaceHistory.py to split based on the # for comments as it removed the date/time from the command, and the idea is that it's stored together. Changing this would require quite a bit of extra work, to restructure how it saves and loads the data.

Reply:
https://github.com/mantidproject/mantid/blob/master/Framework/PythonInterface/plugins/algorithms/OrderWorkspaceHistory.py#L87 is exactly the code I have a problem with, it makes a false assumption about how comments work in Python.

If you use `tokenize` to do this parsing you can robustly extract comments from Python source so having a `#` in a string wouldn't be an issue in the first place.

Reply:
@DanNixon, looking at `tokenize` - it is a bit more involved than I had thought. Do you have any experience using it? I could save us a lot of time by dropping over to you if you do.

Reply:
@keeeto Yes, I can take a look this morning. Do you know of anywhere else this assumption is made apart from `OrderWorkspaceHistory`?